cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,158 @@ def get_sbom_components(self) -> list[dict[str, str]]:
         packages = self._dependencies if self.lockfile_version == 1 else self._packages
         return [{"name": package.name, "version": package.version} for package in packages]
 
+    def get_dependencies_to_download(self) -> Dict[str, Dict[str, Optional[str]]]:
+        """Return a Dict of URL dependencies to download."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return {
+            package.name: {
+                "version": package.version,
+                "url": package.resolved_url,
+                "integrity": package.integrity,
+            }
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def extract_git_info_npm(vcs_url: str) -> Dict[str, str]:
+    """
+    Extract important info from a VCS requirement URL.
+
+    Given a URL such as git+ssh://user@host/namespace/repo.git#9e164b970
+
+    this function will extract:
+    - the "clean" URL: ssh://user@host/namespace/repo.git
+    - the git ref: 9e164b970
+
+    The clean URL and ref can be passed straight to scm.Git to fetch the repo.
+    The host, namespace and repo will be used to construct the file path under deps/npm.
+
+    :param vcs_url: The URL of a VCS requirement, must be valid (have git ref in path)
+    :return: Dict with url, ref, host, namespace and repo keys
+    """
+    # If scheme is git+protocol://, keep only protocol://
+    # Do this before parsing URL, otherwise urllib may not extract URL params
+    if vcs_url.startswith("git+"):
+        vcs_url = vcs_url[len("git+") :]
+
+    ref = vcs_url[-40:]  # Take the last 40 characters (the git ref)
+    clean_url = vcs_url[:-41]  # Drop the last 41 characters ('#' + git ref)
+
+    url = urlparse(clean_url)
+    namespace_repo = url.path.strip("/")
+    if namespace_repo.endswith(".git"):
+        namespace_repo = namespace_repo[: -len(".git")]
+
+    # Everything up to the last '/' is namespace, the rest is repo
+    namespace, _, repo = namespace_repo.rpartition("/")
+
+    vcs_url_info = {
+        "url": clean_url,
+        "ref": ref.lower(),
+        "host": str(url.hostname),
```

Comment:
Casting this to string turns a `None` value into `'None'`. This makes the error on line 244 not be raised.

I tested this function with `"github:kevva/is-positive#97edff6f525f192a3f83cea1944765f769ae2678"`.

(this seems to be a type of VCS URL generated by NPM... see my other comment below)

Reply:
Check fixed